pennylane/optimize/rotosolve.py

```diff
@@ -12,172 +12,427 @@
+                qml.RX(par, wires=i)
 
-    Run the optimization step-by-step for ``n_steps`` steps.
+            for w in dev.wires:
+                qml.RX(layer_par, wires=w)
 
-    >>> cost_rotosolve = []
-    >>> for _ in range(n_steps):
-    ...     cost_rotosolve.append(cost(x))
-    ...     x = opt.step(cost, x)
+            for i, par in enumerate(crot_param):
+                qml.CRY(par, wires=[i, (i+1)%3])
 
-    The optimized values for x should now be stored in ``x`` and steps-vs-cost can be seen by
-    plotting ``cost_rotosel``.
-    """
-    # pylint: disable=too-few-public-methods
+            return qml.expval(qml.PauliZ(0) @ qml.PauliZ(1) @ qml.PauliZ(2))
 
-    def step_and_cost(self, objective_fn, *args, **kwargs):
-        r"""Update args with one step of the optimizer and return the corresponding objective
-        function value prior to the step.
+    This QNode is defined simply by measuring the expectation value of the tensor
+    product of ``PauliZ`` operators on all qubits.
+    It takes three parameters:
 
-        Args:
-            objective_fn (function): The objective function for optimization. It should take a
-                sequence of the values ``*args``  as inputs, and return a single value.
-            *args : variable length argument list containing the initial values of the
-                variables to be optimized over or a single float with the initial value
-            **kwargs : variable length dictionary of keywords for the objective function
+    - ``rot_param`` controls three Pauli rotations with three parameters (one frequency each),
+    - ``layer_par`` feeds into a layer of rotations with a single parameter (three frequencies), and
+    - ``crot_param`` feeds three parameters into three controlled Pauli rotations (two frequencies
+      each).
 
-        Returns:
-            tuple[list [array], float]: the new variable values :math:`x^{(t+1)}` and the objective
-            function output prior to the step.
-            If single arg is provided, list [array] is replaced by array.
-        """
-        x_new = self.step(objective_fn, *args, **kwargs)
+    We also initialize a set of parameters for all these operations, and summarize
+    the numbers of frequencies in ``num_frequencies``.
 
-        return x_new, objective_fn(*args, **kwargs)
+    .. code-block :: python
 
-    def step(self, objective_fn, *args, **kwargs):
-        r"""Update args with one step of the optimizer.
+        init_param = [
+            np.array([0.3, 0.2, 0.67], requires_grad=True),
+            np.array(1.1, requires_grad=True),
+            np.array([-0.2, 0.1, -2.5], requires_grad=True),
+        ]
+
+        num_frequencies = [[1, 1, 1], 3, [2, 2, 2]]
+
+    The keyword argument ``requires_grad`` can be used to determine whether the respective
+    parameter should be optimized or not, following the behaviour of gradient computations and
+    gradient-based optimizers when using Autograd or JAX.
```

Comment:
This is manually implemented in `RotosolveOptimizer`, as no gradients are computed :)
Are you referring to gradient computations with JAX? I don't think I tested this :sweat_smile:

Reply:
Oh, got it!

I don't believe this works for JAX, since
![image](https://user-images.githubusercontent.com/2959003/130091648-d49313a6-5ca4-4667-b0c6-3b8d998e3e80.png)

Reply:
Will adapt the comment! Thank you!